### DISCUSSION TITLE
c++_static运行时需要手动添加-lc++abi

### DISCUSSION THREAD

**Discussion by rocascent** [xmake-io/xmake#5073] [Q&A] [answered]
# 系统
WSL ArchLinux
# 问题
```lua
set_toolchains('clang')
set_runtimes("c++_static")
```
添加了上面的设置之后，需要额外添加`add_ldflags("-lc++abi")`，否则链接时链接器找不到符号。

是否可以通过某种方式自动连接`libc++abi`?

**Comment by waruqi** [comment]
需要改进这里 https://github.com/xmake-io/xmake/blob/3e9e12adad6b63334d25e3861405246afce948c9/xmake/modules/core/tools/clang.lua#L305

你可以尝试提个 pr 过来。

@Arthapz Any idea about `-lc++abi`?

**Reply by Arthapz** [reply]
I don't understand chinese :D I need more context

**Reply by waruqi** [reply]
In wsl archlinux, it seems that if we use c++_static, we need to additionally link -lc++abi, otherwise we get missing link symbols, but I'm not sure if that's necessary.

**Reply by rocascent** [reply]
I'm not sure if it's an issue with the system or the compiler settings, but when using clang, the -lc++abi flag is necessary. For instance, `foo()` will be mangled to `_Z9fooPKc()`.

**Reply by waruqi** [reply]
But it works for me without c++abi.

> /usr/bin/clang -c -Qunused-arguments -m64 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -stdlib=libc++ -DNDEBUG -o build/.objs/test/linux/x86_64/release/src/main.cpp.o src/main.cpp
checking for flags (-MMD -MF) ... ok
checking for flags (-fdiagnostics-color=always) ... ok
checking for clang++ ... /usr/bin/clang++
checking for the linker (ld) ... clang++
checking for /usr/bin/clang++ ... ok
checking for flags (-fPIC) ... ok
checking for flags (clang_static_libstdcxx) ... ok
[ 75%]: linking.release test
/usr/bin/clang++ -o build/linux/x86_64/release/test build/.objs/test/linux/x86_64/release/src/main.cpp.o -m64 -s -stdlib=libc++ -static-libstdc++
[100%]: build ok, spent 23.276s

**Comment by waruqi** [comment]
> I'm not sure if it's an issue with the system or the compiler settings, but when using clang, the -lc++abi flag is necessary. For instance, `foo()` will be mangled to `_Z9fooPKc()`.

and it still works for static lib.

```console
ruki@37f873306484:/mnt/xmake/tests/projects/c++/static_library$ xmake -rv
checking for clang ... /usr/bin/clang
checking for the c++ compiler (cxx) ... clang
checking for /usr/bin/clang ... ok
checking for flags (-fPIC) ... ok
checking for flags (-fvisibility-inlines-hidden) ... ok
checking for flags (-O3) ... ok
checking for flags (-DNDEBUG) ... ok
[ 50%]: cache compiling.release src/foo.cpp
/usr/bin/clang -c -Qunused-arguments -m64 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -stdlib=libc++ -DNDEBUG -o build/.objs/foo/linux/x86_64/release/src/foo.cpp.o src/foo.cpp
checking for flags (-MMD -MF) ... ok
checking for flags (-fdiagnostics-color=always) ... ok
checking for the c++ compiler (cxx) ... clang
[ 62%]: cache compiling.release src/main.cpp
/usr/bin/clang -c -Qunused-arguments -m64 -fvisibility=hidden -fvisibility-inlines-hidden -O3 -stdlib=libc++ -DNDEBUG -o build/.objs/demo/linux/x86_64/release/src/main.cpp.o src/main.cpp
checking for ar ... /usr/bin/ar
checking for the static library archiver (ar) ... ar
[ 62%]: archiving.release libfoo.a
/usr/bin/ar -cr build/linux/x86_64/release/libfoo.a build/.objs/foo/linux/x86_64/release/src/foo.cpp.o
checking for clang++ ... /usr/bin/clang++
checking for the linker (ld) ... clang++
checking for /usr/bin/clang++ ... ok
checking for flags (-fPIC) ... ok
checking for flags (clang_static_libstdcxx) ... ok
[ 87%]: linking.release demo
/usr/bin/clang++ -o build/linux/x86_64/release/demo build/.objs/demo/linux/x86_64/release/src/main.cpp.o -m64 -Lbuild/linux/x86_64/release -s -stdlib=libc++ -static-libstdc++ -lfoo
[100%]: build ok, spent 9.066s
ruki@37f873306484:/mnt/xmake/tests/projects/c++/static_library$ nm build/linux/x86_64/release/libfoo.a

foo.cpp.o:
0000000000000000 T add

```

**Reply by rocascent** [reply]
```
❯ clang -v
clang version 17.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/sbin
Found candidate GCC installation: /usr/sbin/../lib/gcc/x86_64-pc-linux-gnu/13.2.1
Found candidate GCC installation: /usr/sbin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1
Selected GCC installation: /usr/sbin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Selected multilib: .;@m64
```
without -lc++abi, linker has error ```/usr/bin/ld: /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/13.2.1/../../../../lib64/libc++.a(future.cpp.o):(.data.rel.ro._ZTINSt3__123__future_error_categoryE+0x0): undefined reference to `vtable for __cxxabiv1::__si_class_type_info```

**Reply by waruqi** [reply, marked as answer]
try this patch. https://github.com/xmake-io/xmake/pull/5083

**Reply by rocascent** [reply]
great!👍

**Comment by rocascent** [comment]
补充：通过添加`-lc++abi`链接的abi是动态库，可能不符合`c++_static`。出现找不到符号的原因可能是Archlinux仓库中的clang是使用gcc构建的，链接器为ld，所以导致不会自动寻找libc++相关库。

**Reply by waruqi** [reply]
所以 到底是应该加 还是不应该加呢？

**Reply by rocascent** [reply]
既然没有人提issue，感觉还是我个人的问题。出了问题可以自己手动调整编译指令，所以不加吧。

**Reply by waruqi** [reply]
我 revert 了

**Reply by rocascent** [reply]
我自己测试下来，用llvm最新源码编译出的clang 19.0.0，默认使用lld编译器的情况下，链接时添加`-static -lc++abi`即可正确连接静态库。

```sh
$ clang++ -o build/linux/x86_64/debug/demo build/.objs/demo/linux/x86_64/debug/src/main.cpp.o -static -lc++abi
$ file build/linux/x86_64/debug/demo
build/linux/x86_64/debug/demo: ELF 64-bit LSB executable, x86-64, version 1 (SYSV), statically linked, for GNU/Linux 4.4.0, with debug_info, not stripped
```

```sh
$ clang++ -o build/linux/x86_64/debug/demo build/.objs/demo/linux/x86_64/debug/src/main.cpp.o
$ file build/linux/x86_64/debug/demo
build/linux/x86_64/debug/demo: ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 4.4.0, with debug_info, not stripped
```

**Reply by rocascent** [reply]
如果设置了libc++静态链接libc++abi，-lc++abi也不需要，只需 -static 即可编译静态链接版本
